src/Compilers/Core/Portable/MetadataReader/MetadataHelpers.cs

```diff
@@ -505,28 +505,53 @@ private static short InferTypeArityFromMetadataName(string emittedTypeName, out
                 return 0;
             }
 
-            // Given a name corresponding to <unmangledName>`<arity>,
-            // extract the arity.
-            string stringRepresentingArity = emittedTypeName.Substring(indexOfManglingChar);
-
-            int arity;
-            bool nonNumericCharFound = !int.TryParse(stringRepresentingArity, NumberStyles.None, CultureInfo.InvariantCulture, out arity);
-
-            if (nonNumericCharFound || arity < 0 || arity > short.MaxValue ||
-                stringRepresentingArity != arity.ToString())
+            // Given a name corresponding to <unmangledName>`<arity>, extract the arity.
+            if (!tryScanArity(emittedTypeName[indexOfManglingChar..], out var arity))
             {
                 suffixStartsAt = -1;
                 return 0;
             }
 
             suffixStartsAt = indexOfManglingChar - 1;
-            return (short)arity;
+            return arity;
+
+            static bool tryScanArity(ReadOnlySpan<char> aritySpan, out short arity)
+            {
+                // Common case: aritySpan is 1 digit between 0-9.  No need for any complex parsing logic in that case.
+                if (aritySpan is [var firstChar and >= '0' and <= '9'])
+                {
+                    arity = (short)(firstChar - '0');
+                    return true;
+                }
+
+#if NET
+                bool nonNumericCharFound = !int.TryParse(aritySpan, NumberStyles.None, CultureInfo.InvariantCulture, out var intArity);
+#else
+                bool nonNumericCharFound = !int.TryParse(aritySpan.ToString(), NumberStyles.None, CultureInfo.InvariantCulture, out var intArity);
```

Comment:
>int.TryParse [](http://example.com/codeflow?start=44&length=12)

Isn't there an Int16.TryParrse?

Reply:
i'm very wary about changing semantics here (even if it might even fix a bug).  unless there are strong guarantees on the exact semantics of both tryparses, i think preserving this is the safest path.